src/controllers/AddAssetsManager.js

```diff
@@ -0,0 +1,232 @@
+  }
+
+  addUnapprovedAsset(assetParameters, request, id) {
+    // add origin from request
+    if (request) assetParameters.origin = request.origin
+    const time = Date.now()
+    const assetData = {
+      id,
+      assetParams: assetParameters,
+      time,
+      status: 'unapproved',
+      type: MESSAGE_TYPE.WATCH_ASSET,
+    }
+
+    this.addAsset(assetData)
+
+    // signal update
+    this.emit('update')
+    return id
+  }
+
+  async approveAsset(assetId) {
+    const assetData = this.getAsset(assetId)
+    const { options, metadata, type } = assetData.assetParams
+    if (assetData.type.toLowerCase() === CONTRACT_TYPE_ERC20) {
+      const { address, symbol, decimals } = options
+      const { network, name } = metadata
+      await this.prefsController.addCustomToken({
+        token_address: address,
+        network,
+        token_symbol: symbol,
+        token_name: name || symbol,
+        decimals,
+      })
+    } else {
+      const { address, id } = options
+      const { network, nftName } = metadata
+      await this.prefsController.addCustomNft({
+        nft_address: address,
+        network,
+        nft_name: nftName,
+        nft_id: id,
+        nft_contract_standard: type,
+      })
+    }
+    this.setAssetStatusApproved(assetId)
+  }
+
+  rejectAsset(assetId) {
+    this._setAssetStatus(assetId, 'rejected')
+  }
+
+  addAsset(asset) {
+    this.assets.push(asset)
+    this._saveAssetList()
+  }
+
+  getAsset(assetId) {
+    return this.assets.find((asset) => asset.id === assetId)
+  }
+
+  setAssetStatusApproved(assetId) {
+    this._setAssetStatus(assetId, 'approved')
+  }
+
+  _setAssetStatus(assetId, status) {
+    const asset = this.getAsset(assetId)
+    if (!asset) {
+      throw new Error(`AddAssetsManager - Asset not found for id: "${assetId}".`)
+    }
+    asset.status = status
+    this._updateAsset(asset)
+    this.emit(`${assetId}:${status}`, asset)
+    if (status === 'rejected' || status === 'approved' || status === 'errored') {
+      this.emit(`${assetId}:finished`, asset)
+    }
+  }
+
+  _updateAsset(existingAsset) {
+    const index = this.assets.findIndex((asset) => existingAsset.id === asset.id)
+    if (index !== -1) {
+      this.assets[index] = existingAsset
+    }
+    this._saveAssetList()
+  }
+
+  _saveAssetList() {
+    const unApprovedAssets = this.getUnapprovedAssets()
+    const unApprovedAssetsCount = Object.keys(unApprovedAssets).length
+    this.store.updateState({ unApprovedAssets, unApprovedAssetsCount })
+  }
+
+  async validateWatchAssetParams(assetParams, web3) {
+    // log.debug('asset params', assetParams)
+    const supported_asset_types = [CONTRACT_TYPE_ERC20, CONTRACT_TYPE_ERC721, CONTRACT_TYPE_ERC1155]
+    if (!supported_asset_types.includes(assetParams.type.toLowerCase())) {
+      throw ethErrors.rpc.invalidParams('Invalid watch asset params: only erc20/erc721/erc1155 asset types are supported.')
+    }
+    const { address } = assetParams.options || {}
```

Comment:
add check to validate nft id in case of erc721 or erc1155. Id is mandatory for nfts